### ISSUE TITLE
MemorySSA in SLP Vectorizer

### ISSUE THREAD

**Issue by preames** [llvm/llvm-project#54256]
This issue exists to motivate and discuss the use of memory SSA inside SLP.  I have an initial patch up for review (https://reviews.llvm.org/D117926), but there were motivation questions raised there.  I'm filing this for the purpose of providing the motivation and discussion of relevant issues for follow through.

I have a motivating example which spends a large amount of time computing memory dependencies in SLP.  Unfortunately, I can not share the example, and my attempts to produce a synthetic reproducer have not fared well.  The key observations from profiling the example are:
* We make a huge number of calls to calculateDependencies.  Here's the output of some additional statics I added downstream.
```
704357 SLP                          - Number of calcDeps actions
         699021 SLP                          - Number of schedule calls
           5598 SLP                          - Number of ReSchedule actions
             59 SLP                          - Number of ReScheduleOnFail actions
          10084 SLP                          - Number of schedule resets
           8523 SLP                          - Number of vector instructions generated
```
* Looking at the profile (via perf), we are bottlenecked by a combination of accessing the AliasCache (the hash for that map is *red* hot), and performing alias queries.
* This makes sense as SLP recomputes all aliasing dependencies for scheduling purposes every time the "scheduling window" changes.  The "scheduling window" changes many, many times on this example due to a combination of a) many vectorizeable trees, and b) trees which straddle long chunks of a basic block.

Taking a close look at the memory dependence handling, I observe that this is largely duplicating the computation of memoryssa through the optimized walker, but invalidated on every change to the scheduling window.

I believe the right thing to do here is to use memory SSA here.  This is a non-trivial task, both because I'm personally fairly new to mssa, and because I keep tripping across additional bits of unexpected complexity (e.g. bugs in the SLP code or bugs in my mental model of memoryssa)

The reasonable objection is that constructing memoryssa for an entire function eagerly when we might not be able to vectorize at all is undesirable.  I agree, but feel the relevant question is at what point does the inherent inefficiency of the current code outweigh the concern about the cost of computing memory SSA.  On this example, the cost of computing memory ssa is approximately 15 ms - as opposed to several seconds on the current code.  I don't yet have a full implementation, so I can not yet do a broader evaluation.  

Now, let me switch focus and argue for incrementalism in the implementation here.  As noted on the original review, the code posted is *slow*.  Very, very slow.  If enabled, simply preserving MSSA takes roughly 10x longer than the old code.

Despite this, I very much want to land this patch, and then work incrementally.  Why?  Because I've investigated the causes of the slowness, and they're sufficiently complex that I really want a) a working baseline, and b) targeted review for each change.

Here's an overview of my findings so far:
* For the insertion of new loads and stores, the renaming bit is the expensive bit.  However, this is not generic insertion. This is specifically insertion of a new instruction which is known to be clobbered by the set of instructions we're vectorizing (and thus about to remove.)  We can exploit that, but doing so locally revealed a silent bug in SLP where the new instruction is not in fact inserted after the current bundle (despite what all the comments say.)
* For the actual scheduling, we need to update the MSSA graph to account for the moving of instructions.  At the moment, this is the hugely slow part of the patch.  However, this update is actually quite limited.  Assuming a correct schedule, this movement can not change the order of any two clobbering defs and reads.  As such, this is not a general move, but instead a highly restricted (and thus optimizeable) one.  It is however, also a very delicate bit of code, which needs careful consideration, tests, and targeted review.  

Now, let me take a moment and address an alternative.  Why can't we simply restructure the existing code to preserve the memory dependencies across schedule window changes?

The short answer is: we can, at the cost of a bunch of complexity.  

The first thing to note is that preserving the memory dependencies requires the *exact same* updates in the exact same places as memoryssa.  This is fundamental.  We might be able to get away with simpler updates (since the data structure is slightly less complicated), but that's a strong "maybe".  

There's two approaches here:
* First, we could construct the memory dependencies for the block up front, and then preserve them through transformation.  This is the simplest of the two options, but has the effect of largely duplicating the functionality MemorySSA provides just specialized for working a block at a time.  
* Second, we could try to preserve the schedule window isolation (e.g. only compute dependencies for things in the schedule window).  This adds a *ton* of complexity.  As in, I haven't yet figured out how to implement it.  I view this option as needing a lot of motivation to justify the complexity, and at the moment, I don't see that motivation at all.  

Personally, I think going with Memory SSA is clearly the best answer.  We might find out it doesn't work for some reason, but we always have that possibility.  If so, we revert a couple changes, and try something else.  I don't see that risk as being a reason not to pursue Memory SSA for SLP in tree.

**Comment by nikic** [comment]
Can you comment on why we can't bail out from vectorization if we need to perform too many alias queries? Similarly to what MemorySSA would do internally? For pathological cases, a bailout is quite often preferable over any complex solutions.

**Comment by preames** [comment]
Well, it turns out I was wrong on memory ssa being a good idea in SLP.  I'm about the revert the change that landed, but let me summarize the piece that tripped me up to save others work in the future.

The key thing we use memory aliasing analysis for in SLP is to identify dependency edges for scheduling.  As a brief review, a dependence edge exists any two memory accesses where at least one is a write.  Typically, we categorize them as RAW (read after write), WAW (write after write) and WAR (write after read).

My goal with this patch series was to use memory ssa to efficiently identify the dependence edges, and thus be able to entirely delete the aliasing walk done by SLP.

I had thought I had a working implementation thereof, but when I went to clean it up and post for review, I realized I'd mishandled in a rather important way the WAR case.  After some thought - and an email exchange with Alina - I've concluded that MSSA does not give us an efficient way to identify write-after-read dependence edges.  We can re-implement basically the same algorithm over MSSA, but that's the best we can do.  

MSSA does give us an efficient way (using optimized edges in memory ssa, and some admittedly questionable invariant assumptions) to handle the read-after-write and write-after-write cases.  

The result is that while we can use mssa for an optimization of some cases inside calculateDependencies, we *can't* replace the existing system in it's entirety.  As a result, none of the code simplifications mentioned previously are actually possible.

Given that, the question becomes is adding additional complexity for a memory-ssa RAW and WAW detection worth while?  

My personal answer at the moment is "no".  It was one situation when doing all the mssa work was expected to remove almost as much complexity as it added, but now facing the realization it's a strict addition of complexity for something which is still a bit speculative in its benefit at all, I've decided this is not worth pursuing.

Someone could reasonable revisit this decision in the future, but in the near term I'm going to discontinue work and revert the initial patch which has already landed.
